lib/Target/AArch64/AArch64RegisterInfo.cpp

```diff
@@ -76,9 +76,11 @@ AArch64RegisterInfo::getCalleeSavedRegs(const MachineFunction *MF) const {
   assert(MF && "Invalid MachineFunction pointer.");
 
   if (MF->getFunction().getCallingConv() == CallingConv::GHC)
-    // GHC set of callee saved regs is empty as all those regs are
-    // used for passing STG regs around
-    return CSR_AArch64_NoRegs_SaveList;
+    // By default LLVM doesn't save any regs for GHC.
+    // This means we'll clobber LR on arm64 if we make non-tail calls (e.g. L2 syscall)
+    // So we should save the normal set of registers instead, because we move
+    // between GHC and normal calling convention.
```

Comment:
Isn't saving only LR enough? Avoiding big pointless prologues was one of the strong points of GHC CC on x86.